src/com/google/enterprise/adaptor/database/ResponseGenerator.java

```diff
@@ -386,19 +401,98 @@ public void generateResponse(ResultSet rs, Response resp)
-        // SerialBlob is adamant about not supporting free
-        try {
+      switch (columnType) {
+        case Types.LONGVARCHAR:
+          try (Reader reader = rs.getCharacterStream(index);
+              Writer writer = new OutputStreamWriter(out, UTF_8)) {
+            if (reader != null) {
+              copy(reader, writer);
+            }
+          }
+          break;
+        case Types.VARBINARY:
+        case Types.LONGVARBINARY:
+          try (InputStream in = rs.getBinaryStream(index)) {
+            if (in != null) {
+              IOHelper.copyStream(in, out);
+            }
+          }
+          break;
+        case Types.CLOB:
+          Clob clob = rs.getClob(index);
+          if (clob != null) {
+            try (Reader reader = clob.getCharacterStream();
+                Writer writer = new OutputStreamWriter(out, UTF_8)) {
+              copy(reader, writer);
+            }
+          }
+          clob.free();
+          break;
+        case Types.NCLOB:
+          NClob nclob = rs.getNClob(index);
+          if (nclob != null) {
+            try (Reader reader = nclob.getCharacterStream();
+                Writer writer = new OutputStreamWriter(out, UTF_8)) {
+              if (reader != null) {
+                copy(reader, writer);
+              }
+            }
+          }
+          nclob.free();
+          break;
+        case Types.BLOB:
+          Blob blob = rs.getBlob(index);
+          if (blob != null) {
+            try (InputStream in = blob.getBinaryStream()) {
+              IOHelper.copyStream(in, out);
+            }
+          }
           blob.free();
-        } catch (java.sql.SQLFeatureNotSupportedException | 
-            UnsupportedOperationException unsupported) {
-          // let JVM garbage collection deal with it
-        }
+          break;
+        case Types.NCHAR:
+        case Types.NVARCHAR:
+        case Types.LONGNVARCHAR:
+          try (Reader reader = rs.getNCharacterStream(index);
+              Writer writer = new OutputStreamWriter(out, UTF_8)) {
+            if (reader != null) {
+              copy(reader, writer);
+            }
+          }
+          break;
+        case Types.SMALLINT:
+        case Types.TINYINT:
+        case Types.FLOAT:
+        case Types.DECIMAL:
+        case Types.NUMERIC:
+        case Types.DOUBLE:
+        case Types.INTEGER:
+        case Types.CHAR:
+        case Types.VARCHAR:
+          String str = rs.getString(index);
+          if (str != null) {
+            out.write(str.getBytes(UTF_8));
+          }
+          break;
+        default:
+          //TODO(srinivas): handle BFILE
+          log.log(Level.FINEST, "Column type not handled: {0}", columnType);
+          break;
+      }
+    }
+
+    private void copy(Reader reader, Writer writer) throws IOException {
+      char buffer[] = new char[1024];
+      int len;
+      while ((len = reader.read(buffer, 0, 1024)) != -1) {
```

Comment:
s/, 0, 1024//

This is needed on the write, but on reads we can just fill the buffer.